banyand/stream/service.go

```diff
@@ -94,101 +101,266 @@ func (s *service) PreRun() error {
+			s.l.Error().Err(innerErr).Msg("fail to initialize stream")
 		}
-		id := formatStreamID(sm.name, sm.group)
-		s.schemaMap[id] = sm
-		s.l.Info().Str("id", id).Msg("initialize stream")
 	}
-	s.writeListener = setUpWriteCallback(s.l, s.schemaMap)
+	s.writeListener = setUpWriteCallback(s.l, &s.schemaMap)
 	return err
 }
 
 func (s *service) Serve() error {
-	t := time.Now()
-	now := time.Now().UnixNano()
-	nowBp := timestamppb.New(t)
-	for _, sMeta := range s.schemaMap {
-		locator := make([]*databasev1.EntityEvent_TagLocator, 0, len(sMeta.entityLocator))
-		for _, tagLocator := range sMeta.entityLocator {
-			locator = append(locator, &databasev1.EntityEvent_TagLocator{
-				FamilyOffset: uint32(tagLocator.FamilyOffset),
-				TagOffset:    uint32(tagLocator.TagOffset),
+	s.schemaMap.Range(func(key, value interface{}) bool {
+		s.l.Debug().Str("streamID", key.(string)).Msg("serve stream")
+		s.serveStream(value.(*stream))
+		return true
+	})
+
+	errWrite := s.pipeline.Subscribe(data.TopicStreamWrite, s.writeListener)
+	if errWrite != nil {
+		return errWrite
+	}
+
+	s.stopCh2 = make(chan struct{})
+	// run a serial reconciler
+	go s.reconcile()
+
+	s.metadata.StreamRegistry().RegisterHandler(schema.KindStream|schema.KindIndexRuleBinding|schema.KindIndexRule, s)
+
+	s.stopCh = make(chan struct{})
+	<-s.stopCh
+
+	return nil
+}
+
+func (s *service) reconcile() {
+	for {
+		select {
+		case streamEntity := <-s.updateCh:
+			s.reloadStream(streamEntity)
+		case metadataEntity := <-s.deleteCh:
+			s.removeStream(metadataEntity)
+		case <-s.stopCh2:
+			return
+		}
+	}
+}
+
+func (s *service) OnAddOrUpdate(m schema.Metadata) {
+	switch m.Kind {
+	case schema.KindStream:
+		s.updateCh <- m.Spec.(*databasev1.Stream)
+	case schema.KindIndexRuleBinding:
+		if m.Spec.(*databasev1.IndexRuleBinding).GetSubject().Catalog == commonv1.Catalog_CATALOG_STREAM {
+			stm, err := s.metadata.StreamRegistry().GetStream(context.TODO(), &commonv1.Metadata{
+				Name:  m.Name,
+				Group: m.Group,
 			})
+			if err != nil {
+				s.l.Error().Err(err).Msg("fail to get subject")
+				return
+			}
+			s.updateCh <- stm
 		}
-		_, err := s.repo.Publish(event.StreamTopicEntityEvent, bus.NewMessage(bus.MessageID(now), &databasev1.EntityEvent{
-			Subject: &commonv1.Metadata{
-				Name:  sMeta.name,
-				Group: sMeta.group,
+	case schema.KindIndexRule:
+		subjects, err := s.metadata.Subjects(context.TODO(), m.Spec.(*databasev1.IndexRule), commonv1.Catalog_CATALOG_STREAM)
+		if err != nil {
+			s.l.Error().Err(err).Msg("fail to get subjects(stream)")
+			return
+		}
+		for _, sub := range subjects {
+			s.updateCh <- sub.(*databasev1.Stream)
+		}
+	default:
+	}
+}
+
+func (s *service) OnDelete(m schema.Metadata) {
+	switch m.Kind {
+	case schema.KindStream:
+		s.deleteCh <- &commonv1.Metadata{
+			Name:  m.Name,
+			Group: m.Group,
+		}
+	case schema.KindIndexRuleBinding:
+		// TODO: we only have metadata(group and name) here, what shall we remove?
```

Comment:
just ignore them

Reply:
Now, I use an option in etcd client to get the previous value while deleting keys. So it is possible to know later which entity has been deleted.

Reply:
nice